src/hotspot/share/opto/superword.cpp

```diff
@@ -1214,110 +1186,136 @@ bool SuperWord::follow_def_uses(Node_List* p) {
-    // Arrange all sub components by the major component
-    if (p2 != nullptr) {
-      for (uint j = 1; j < p->size(); j++) {
-        Node* d1 = p->at(j);
-        Node* u1 = p2->at(j);
-        opnd_positions_match(s1, t1, d1, u1);
-      }
-    }
+    // Find pair (use1, use2)
+    Node* use2 = _pairset.get_right_or_null_for(use1);
+    if (use2 == nullptr) { break; }
+
+    order_inputs_of_uses_to_match_def_pair(def1, def2, use1, use2);
   }
 }
 
-//---------------------------opnd_positions_match-------------------------
-// Is the use of d1 in u1 at the same operand position as d2 in u2?
-bool SuperWord::opnd_positions_match(Node* d1, Node* u1, Node* d2, Node* u2) {
-  // check reductions to see if they are marshalled to represent the reduction
-  // operator in a specified opnd
-  if (is_marked_reduction(u1) && is_marked_reduction(u2)) {
-    // ensure reductions have phis and reduction definitions feeding the 1st operand
-    Node* first = u1->in(2);
+// For a def-pair (def1. def2), and their use-nodes (use1, use2):
+// ensure that the input order of (use1, use2) matches the order of (def1, def2).
+//
+// We have different cases:
+//
+// 1. Reduction (use1, use2): must always reduce left-to-right. Make sure that we have pattern:
+//
+//    phi/reduction x1  phi/reduction x2                    phi/reduction x1
+//                | |               | |    and hopefully:               | |
+//                use1              use2                                use1 x2
+//                                                                         | |
+//                                                                         use2
+//
+// 2: Inputs of (use1, use2) already match (def1, def2), i.e. for all input indices i:
+//
+//    use1->in(i) == def1 || use2->in(def2)   ->    use1->in(i) == def1 && use2->in(def2)
+//
+// 3: Add/Mul (use1, use2): we can try to swap edges:
+//
+//     def1 x1   x2 def2           def1 x1   def2 x2
+//        | |     | |       ==>       | |       | |
+//        use1    use2                use1      use2
+//
+// 4: MulAddS2I (use1, use2): we can try to swap edges:
+//
+//    (x1 * x2) + (x3 * x4)    ==>  4.a: (x2 * x1) + (x4 * x3)
+//                                  4.b: (x4 * x3) + (x2 * x1)
+//                                  4.c: (x3 * x4) + (x1 * x2)
+//
+//    Note: MulAddS2I with its 4 inputs is too complicated, if there is any mismatch, we always
+//          return PairOrderStatus::Unknown.
+//          Therefore, extend_pairset_with_more_pairs_by_following_use cannot extend to MulAddS2I,
+//          but there is a chance that extend_pairset_with_more_pairs_by_following_def can do it.
+//
+SuperWord::PairOrderStatus SuperWord::order_inputs_of_uses_to_match_def_pair(Node* def1, Node* def2, Node* use1, Node* use2) {
+  assert(_pairset.has_pair(def1, def2), "(def1, def2) must be a pair");
+
+  // 1. Reduction
+  if (is_marked_reduction(use1) && is_marked_reduction(use2)) {
+    Node* first = use1->in(2);
```

Comment:
I agree, it already was like that, and it disturbed it me too. I thought I'd leave it since I'll probably overhaul this whole code soon anyway. But I'll fix it already since it disturbs you too.